core/src/main/java/com/linecorp/armeria/client/SessionProtocolNegotiationCache.java

```diff
@@ -61,17 +64,27 @@ protected boolean removeEldestEntry(Entry<String, CacheEntry> eldest) {
      * Returns {@code true} if the specified {@link Endpoint} is known to have no support for
      * the specified {@link SessionProtocol}.
      */
-    public static boolean isUnsupported(Endpoint endpointWithPort, SessionProtocol protocol) {
-        requireNonNull(endpointWithPort, "endpointWithPort");
-        checkArgument(endpointWithPort.hasPort(), "endpointWithPort must have a port.");
-        // It's okay to create the key using endpointWithPort.host():
-        // - If the endpoint has an IP address without host name, the IP address is used in the key to store
-        //   and retrieve the value.
-        // - If the endpoint has an IP address with host name, the host name is used in the key to store
-        //   and retrieve the value.
-        // - If the endpoint has host name only, the host name is used in the key to store and retrieve
-        //   the value.
-        final String key = key(endpointWithPort.host(), endpointWithPort.port());
+    public static boolean isUnsupported(Endpoint endpoint, SessionProtocol protocol) {
+        requireNonNull(endpoint, "endpoint");
+        final String key;
+        if (endpoint.isDomainSocket()) {
+            try {
+                key = "unix:" + URLDecoder.decode(endpoint.host(), "UTF-8");
```

Comment:
Could you remove `unix:` prefix? `endpoint.host()` is created from `DomainSocketAddress.authority()` that seems to start with `unix%3A`.

I couldn't test Unix domain socket on my macOS, but I speculate the key is `unix:unix:/path/to.sock`

Reply:
You're correct. Added some tests to make sure this never happens again.